### PR TITLE
FOV QC

### PR DESCRIPTION
**Pull request by patrickjdanaher**
Adding a script providing a tool for FOV QC. See the new blog post for the general idea, then the code folder for the code itself plus associated files.

### CODE REVIEW

File: code/FOV QC/README.md [hunk 1]
```diff
@@ -0,0 +1 @@
+For a full explanation of the method implemented here, see [our blog post on the subject](../blog/FOV%20QC.md).
```
Comment (crwilliams11) [inline code/FOV QC/README.md:1]: Link doesn't work. Should "/code/blog" be just "/blog"?

File: code/FOV QC.R [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    }
+  }
+  return(bitmat)
+}
+
+#' Convert gene counts to bit counts
+#' @param mat Counts matrix
+#' @param genes Vector of gene names (omit negprobes and falsecodes)
+#' @param barcodes Vector of gene barcodes, aligned to genes
+genes2bits <- function(mat, genes, barcodes) {
```
Comment (crwilliams11) [inline code/FOV QC.R]: This is a useful helper function to have, but is not called elsewhere in utils or referenced in the vignette. Suggestion is to simplify cellxgene2squarexbit by calling this function, as the two are highly repetitive.

File: code/FOV QC.R [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+  return(list(gridid = grid, gridfov = gridfov))
+}
+
+
+#' Convery cell x gene matrix to grid square * bit matrix
+#' @param counts Counts matrix
+#' @param grid Vector assigning cells to squares, aligned to rows of counts
+#' @param genes Vector of gene names (omit negprobes and falsecodes)
+#' @param barcodes Vector of gene barcodes, aligned to genes
+cellxgene2squarexbit <- function(counts, grid, genes, barcodes) {
```
Comment (crwilliams11) [inline code/FOV QC.R]: Can you call genes2bits here to simplify?

File: code/FOV QC.R [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+  comparators <- getNearestNeighborsByFOV(x = bitcounts, 
+                                          fov = gridinfo$gridfov, 
+                                          n_neighbors = 10)
+  
+  # get expected:
+  yhat <- t(sapply(1:nrow(bitcounts), function(i) {
+    colMeans(bitcounts[comparators[i, ], ])
+  }))
+  # get resids:
+  resid = log2((bitcounts + 5) / (yhat + 5))
```
Comment (crwilliams11) [inline code/FOV QC.R]: Why add 5 here?

File: blog/FOV QC.md [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+log2(fold-changes) (red squares) appear to follow spatially smooth biology and not the 
+sharp FOV borders, suggesting we needn't worry about technical artifacts in this FOV. 
+
+Our tool summarize our output across FOVs x barcode bits with plots like the below:
+
+![image](https://github.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space/assets/4357938/1faead6b-1111-4133-9c4d-555ebde70c6b)
+
+In this example, 2 barcode bits from reporter cycle 12 were flagged, as was one bit from reporter cycle 18. 
+Because all known artifacts impact *reporter cycles*, not the individual colors within them, we only flag FOVs
+in which at least two bits/colors from a single reporter cycle appear anomalous. This rule helps avoid flagging
```
Comment (crwilliams11) [inline blog/FOV QC.md:105]: In this case, FOV 19 would be flagged since there are 2 flagged bits from reporter cycle 12 whereas FOV 18 would not be flagged since there was only one bit flagged in reporter cycle 18.

File: code/FOV QC/FOV QC utils.R [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+}
+
+
+#' Heatmap of estimated bit bias across FOVs
+#' 
+#' @param res Results object created by runFOVQC
+#' @return Draws a heatmap
+#' @export
+FOVEffectsHeatmap <- function(res) {
+  pheatmap::pheatmap(res$fovstats$bias * (res$fovstats$flag),
```
Comment (crwilliams11) [inline code/FOV QC/FOV QC utils.R:266]: To make it easier to read, add y-axis title 'FOV' and x-axis title 'Reporter cycle'. Remove 'reportercycle' from x-axis labels since it's repetitive and makes the axis ticks busier

File: code/FOV QC/FOV QC utils.R [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+  })
+}
+
+
+#' Heatmap of estimated bit bias across FOVs
+#' 
+#' @param res Results object created by runFOVQC
+#' @return Draws a heatmap
+#' @export
+FOVEffectsHeatmap <- function(res) {
```
Comment (crwilliams11) [inline code/FOV QC/FOV QC utils.R:265]: Heatmap shows an empty white square for every FOV and bit that was evaluated but not flagged, in some cases leading to lots of white space. Can you add an option to only show FOVs / bits where at least one was flagged for bias? This would remove a lot of white space and make it easier to read, especially for samples with many FOVs

File: code/FOV QC/FOV QC utils.R [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+#' Spatial plots of FOV effects:
+#' 
+#' @param res Results object created by runFOVQC
+#' @param outdir Directory where png plots are printed
+#' @param bits Which bits to plot. Defaults to "flagged_reportercycles" bits, but can also plot "all" or "flagged_bits".
+#' @param plotwidth Width in inches of png plots
+#' @param plotheight Height in inches of png plots
+#' @return For each bit, draws a plot of estimated FOV effects
+#' @export
+FOVEffectsSpatialPlots <- function(res, outdir = NULL, bits = "flagged_reportercycles", plotwidth = NULL, plotheight = NULL) {
```
Comment (crwilliams11) [inline code/FOV QC/FOV QC utils.R:128]: Enhancement suggestion: allow to specify single bit or list of bits with the argument `bits`, in addition to the three listed options. Useful for quick plotting in IDE

File: code/FOV QC/FOV QC vignette.R [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+head(fov)
+# (Note : the above data objects can easily be extracted from a Giotto or Seurat or tileDB object. 
+# See the docs for whichever system you're using for how to extract count matrices and metadata columns.
+# Note that this code expects a count matrix with cells in rows & genes in columns, and 
+# some toolkits store a transposed version of this matrix. In this case, use Matrix::t() to transpose
+# your sparse counts matrix.)
+
+
+#### run QC pipeline -------------------
+res <- runFOVQC(counts = counts, xy = xy, fov = fov, barcodemap = barcodemap, max_prop_loss = 0.2)
```
Comment (crwilliams11) [inline code/FOV QC/FOV QC vignette.R]: Default max loss is 0.3. Here might be a good place to add some explanation or guidance about how to set this appropriately.

File: code/FOV QC/FOV QC utils.R [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+#' @param counts Raw counts matrix, cells in rows, genes in columns. Can be sparse or standard format.
+#' @param xy 2-column matrix of cells' xy positions, aligned to rows of counts.
+#' @param fov Vector of cells' FOV IDs, aligned to rows of counts.
+#' @param barcodemap Data frame with two columns: "gene" and "barcode". Download the barcodemap for your panel
+#' from https://github.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space/tree/Main/code/FOV%20QC.
+#' @param max_prop_loss Maximum loss of efficiency allowed for any bit. E.g., a value of "0.3" means all FOVs with bias <log2(1 - 0.3) will be flagged.
+#' @export
+runFOVQC <- function(counts, xy, fov, barcodemap, max_prop_loss = 0.3) {
+  
+  if ((max_prop_loss > 1) | (max_prop_loss < 0)) {
```
Comment (crwilliams11) [inline code/FOV QC/FOV QC utils.R:25]: Are 1 and 0 valid values here? Should these be `>=` and `<=` ?

File: code/FOV QC/FOV QC utils.R [hunk 11]
```diff
@@ -0,0 +1,297 @@
...
+#' to similar squares in other FOVs, and look for bardode bits whose genes are losing signal.
+#' Recommended to use this function for one tissue/slide at a time, not across multiple tissues.
+#' @param counts Raw counts matrix, cells in rows, genes in columns. Can be sparse or standard format.
+#' @param xy 2-column matrix of cells' xy positions, aligned to rows of counts.
+#' @param fov Vector of cells' FOV IDs, aligned to rows of counts.
+#' @param barcodemap Data frame with two columns: "gene" and "barcode". Download the barcodemap for your panel
+#' from https://github.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space/tree/Main/code/FOV%20QC.
+#' @param max_prop_loss Maximum loss of efficiency allowed for any bit. E.g., a value of "0.3" means all FOVs with bias <log2(1 - 0.3) will be flagged.
+#' @export
+runFOVQC <- function(counts, xy, fov, barcodemap, max_prop_loss = 0.3) {
```
Comment (crwilliams11) [inline code/FOV QC/FOV QC utils.R]: Not all FOVs get evaluated due to sparse cells. Print out a list / add to results object a list of FOVs that did not get evaluated.

File: code/FOV QC/FOV QC utils.R (hunk 8 above)
Comment (crwilliams11) [reply code/FOV QC/FOV QC utils.R:128]: Additional enhancement: Add option to print FOV number on top of grid, similar to what's shown in blog

File: code/FOV QC/FOV QC utils.R [hunk 12]
```diff
@@ -0,0 +1,297 @@
...
+  
+  if ((max_prop_loss > 1) | (max_prop_loss < 0)) {
+    stop("max_prop_loss must fall in range of 0-1.")
+  }
+  fov <- as.character(fov)
+  ## create a matrix of barcode bit expression over sub-FOV grids:
+  # define grids, get per-square gene expression:
+  gridinfo <- makeGrid(mat = counts, xy = xy, fov = fov, squares_per_fov = 49, min_cells_per_square = 25)
```
Comment (crwilliams11) [inline code/FOV QC/FOV QC utils.R]: Why not expose `squares_per_fov` and `min_cells_per_square` as arguments to `runFOVQC`, setting 49 and 25 as the defaults?

File: blog/FOV QC.md [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+in other FOVs, with "similar" being based on the square's expression profile. (We also only accept one 
+neighbor per other FOV.)
+Then, for each barcode bit, we take the genes using the bit, and we contrast their 
+expression in the square vs. in the average of the 10 most similar squares elsewhere. 
+For a given FOV and barcode bit, this gives us 49 contrasts. 
+When an FOV's grid squares consistently underexpress the relevant gene set, we flag the FOV.
+
+Below we demonstrate this approach, looking at a tissue with particularly dramatic FOV effects. 
+
+![image](https://github.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space/assets/4357938/76c95e5d-9bef-4bea-9571-84b37cf9b988)
```
Comment (crwilliams11) [inline blog/FOV QC.md:84]: Figure title on the left says 'by sub-FOV region' but plot appears to show expression by cell. Change either the title or what's plotted -- I'd recommend the latter for easier comparison to plot on right.

File: blog/FOV QC.md [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+FOV 22 now stands out as having perhaps increased expression of the bit, but the high
+log2(fold-changes) (red squares) appear to follow spatially smooth biology and not the 
+sharp FOV borders, suggesting we needn't worry about technical artifacts in this FOV. 
+
+Our tool summarize our output across FOVs x barcode bits with plots like the below:
+
+![image](https://github.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space/assets/4357938/1faead6b-1111-4133-9c4d-555ebde70c6b)
+
+In this example, 2 barcode bits from reporter cycle 12 were flagged, as was one bit from reporter cycle 18. 
+Because all known artifacts impact *reporter cycles*, not the individual colors within them, we only flag FOVs
```
Comment (crwilliams11) [inline blog/FOV QC.md:104]: Reword. Earlier we also refer to autofluorescence as driving artifacts, which impact individual colors rather than reporter cycles.

File: blog/FOV QC.md (hunk 5 above)
Comment (crwilliams11) [reply blog/FOV QC.md:105]: Add the above or similar to help w/ explanation

File: code/FOV QC/FOV QC utils.R [hunk 15]
```diff
@@ -0,0 +1,297 @@
...
+    # none from own FOV:
+    tempneighbors[fov[topneighbors[i, ]] == fov[i]] <- NA
+    return(tempneighbors[!is.na(tempneighbors)][1:n_neighbors])
+  }))
+  
+  return(reducedneighbors)
+}
+
+
+#' Get gridded expression within FOVs
```
Comment (crwilliams11) [inline code/FOV QC/FOV QC utils.R]: This makes grids but doesn't get expression per grid. Change description.

File: code/FOV QC/FOV QC utils.R [hunk 16]
```diff
@@ -0,0 +1,297 @@
...
+#' @param xy 2-column matrix of xy locations
+#' @param fov Vector of FOV IDs
+#' @param squares_per_fov Number of squares to break fov into
+#' @param min_cells_per_square Fewest cells to keep a square
+#' @return A list: a matrix of expression in grid squares, a vector giving the grid square assignment of each cell.   
+makeGrid <- function(mat, xy, fov, squares_per_fov = 49, min_cells_per_square = 25) {
+  
+  # make grids:
+  ncuts <- floor(sqrt(squares_per_fov))
+  grid <- rep(NA, nrow(mat))
```
Comment (crwilliams11) [inline code/FOV QC/FOV QC utils.R]: Can get number of cells from `xy` and then don't need to provide expression matrix to this function

File: code/FOV QC/FOV QC utils.R [hunk 17]
```diff
@@ -0,0 +1,297 @@
...
+
+
+#' Get gridded expression within FOVs
+#' 
+#' @param mat Expression matrix, linear-scale
+#' @param xy 2-column matrix of xy locations
+#' @param fov Vector of FOV IDs
+#' @param squares_per_fov Number of squares to break fov into
+#' @param min_cells_per_square Fewest cells to keep a square
+#' @return A list: a matrix of expression in grid squares, a vector giving the grid square assignment of each cell.
```
Comment (crwilliams11) [inline code/FOV QC/FOV QC utils.R]: Revise this, no expression matrix is returned

File: code/FOV QC/FOV QC utils.R [hunk 18]
```diff
@@ -0,0 +1,297 @@
...
+  
+  # clean up low-data grids:
+  toosmallgridids <- names(which(table(grid) < min_cells_per_square))
+  grid[is.element(grid, toosmallgridids)] <- NA
+  
+  return(list(gridid = grid, gridfov = gridfov))
+}
+
+
+#' Convert cell x gene matrix to grid square * bit matrix
```
Comment (crwilliams11) [inline code/FOV QC/FOV QC utils.R:331]: add `@return`

File: code/FOV QC/FOV QC utils.R [hunk 19]
```diff
@@ -0,0 +1,297 @@
...
+      tempgenes <- intersect(tempgenes, colnames(counts))
+      temptotal <- Matrix::rowSums(counts[, tempgenes, drop = FALSE])
+      tempsquaretotal <- by(temptotal, grid, mean)
+      bitmat[names(tempsquaretotal), paste0("reportercycle", i, col)] <- tempsquaretotal
+    }
+  }
+  return(bitmat)
+}
+
+#' Convert gene counts to bit counts
```
Comment (crwilliams11) [inline code/FOV QC/FOV QC utils.R]: add `@return`

**Review by crwilliams11 — CHANGES_REQUESTED** [review]
Comments throughout. Most should be addressed before merging, though some are labeled as potential enhancements and I'd be open to adding others to future enhancements if needed.

File: code/FOV QC.R (hunk 4 above)
Comment (patrickjdanaher) [reply code/FOV QC.R]: I didn't want it going crazy for bits with very low total expression. I.e. I didn't want log(2/1) or log(0/1). The choice of 5 is arbitrary, but I think in the right ballpark.

File: blog/FOV QC.md (hunk 5 above)
Comment (patrickjdanaher) [reply blog/FOV QC.md:105]: good idea. done.

File: code/FOV QC/FOV QC utils.R (hunk 16 above)
Comment (patrickjdanaher) [reply code/FOV QC/FOV QC utils.R]: good catch. fixed.

File: code/FOV QC/FOV QC utils.R (hunk 17 above)
Comment (patrickjdanaher) [reply code/FOV QC/FOV QC utils.R]: corrected

File: code/FOV QC/FOV QC utils.R (hunk 18 above)
Comment (patrickjdanaher) [reply code/FOV QC/FOV QC utils.R:331]: done

File: code/FOV QC/FOV QC utils.R (hunk 19 above)
Comment (patrickjdanaher) [reply code/FOV QC/FOV QC utils.R]: done

File: code/FOV QC/FOV QC utils.R (hunk 15 above)
Comment (patrickjdanaher) [reply code/FOV QC/FOV QC utils.R]: done

File: code/FOV QC.R (hunk 3 above)
Comment (patrickjdanaher) [reply code/FOV QC.R]: I used to do that, but I save a ton of memory / computation time under the current setup - I never need to create a large dense matrix this way.

File: code/FOV QC.R (hunk 2 above)
Comment (patrickjdanaher) [reply code/FOV QC.R]: Good catch. I think I'll just delete this one, since the function cellxgene2squarexbit can't be reduced without costing computation time.

File: code/FOV QC/README.md (hunk 1 above)
Comment (patrickjdanaher) [reply code/FOV QC/README.md:1]: I believe it works now.

File: code/FOV QC.R (hunk 2 above)
Comment (crwilliams11) [reply code/FOV QC.R]: Sounds good to me

File: code/FOV QC/README.md (hunk 1 above)
Comment (crwilliams11) [reply code/FOV QC/README.md:1]: Still not working for me, but this link will need to be updated to location on website blog before merging into main anyway

File: code/FOV QC/FOV QC vignette.R (hunk 9 above)
Comment (crwilliams11) [reply code/FOV QC/FOV QC vignette.R]: Has been added.

**Comment by eveilyeverafter** [comment]
Can this be closed?
